aptos-move/block-executor/src/executor.rs

```diff
@@ -207,27 +252,42 @@ where
         versioned_cache: &MVHashMap<T::Key, T::Value, ExecutableTestType>,
         scheduler: &Scheduler,
         base_view: &S,
-        committing: bool,
+        role: CommitRole,
     ) {
         // Make executor for each task. TODO: fast concurrent executor.
         let init_timer = VM_INIT_SECONDS.start_timer();
         let executor = E::init(*executor_arguments);
         drop(init_timer);
 
+        let committing = matches!(role, CommitRole::Coordinator(_, _));
+
         let _timer = WORK_WITH_TASK_SECONDS.start_timer();
         let mut scheduler_task = SchedulerTask::NoTask;
         loop {
-            // Only one thread try_commit to avoid contention.
-            if committing {
-                // Keep committing txns until there is no more that can be committed now.
-                while let Some(txn_idx) = scheduler.try_commit() {
-                    if txn_idx as usize + 1 == block.len() {
-                        // Committed the last transaction / everything.
-                        scheduler_task = SchedulerTask::Done;
-                        break;
+            // Only one thread does try_commit to avoid contention.
+            match &role {
+                CommitRole::Coordinator(post_commit_txs, mut idx) => {
+                    while let Some(txn_idx) = scheduler.try_commit() {
+                        post_commit_txs[idx]
+                            .send(txn_idx)
+                            .expect("Worker must be available");
+                        // Iterate round robin over workers to do commit_hook.
+                        idx = (idx + 1) % post_commit_txs.len();
+
+                        if txn_idx as usize + 1 == block.len() {
+                            // Committed the last transaction / everything.
+                            scheduler_task = SchedulerTask::Done;
+                            break;
```

Comment:
this might be incorrect if there's only one worker, (it breaks without draining the queue), we don't spawn one worker for execution today but seems like a bomb

Reply:
We have an assert for it? I can add a comment to the assert that it wouldn't actually work? (and it would be not too hard to fix if it was ever needed somehow - just calling hook itself.).